core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java

```diff
@@ -299,6 +308,13 @@ private boolean isEndpointGatewayEnabled(EnvironmentNetworkResponse network) {
         return Optional.ofNullable(knoxTargetGroup);
     }
 
+    private void validateKnoxForAzure(Set<InstanceGroup> knoxGatewayInstanceGroups, String cloudPlatform) {
```

Comment:
this is not a validation as it modifies the list (therefore it should return with the list instead of modifying the parameter)
shouldn't this throw an error an prevent provisioning instead of silently going on?

Reply:
I'll fix the method name.

As for the error, we've been trying to allow placement of Knox in several Instance groups since that _should_ be a valid configuration.

I think we have two options:
* Throw an error if we can't provision -> prevent Knox from living in multiple instance groups while we load balance
* Don't throw an error, but also don't provision a load balancer -> Knox in multiple instance groups is still valid, but we quietly do unexpected things.

We can't have Knox live in multiple instance groups and route to them using an Azure LB because we're still using the `basic` SKU. If we could switch to the `standard` SKU, this check wouldn't even exist.

Reply:
Actually, I took a look at this and refactored it:
* Remove the `validateKnoxForAzure` method
* Move the check up, and make it a branch of an `if/else if` block

We still skip provisioning load balancers, instead of throwing an error to stop everything altogether.

```diff
diff --git a/core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java b/core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
index 1f1c31a032..81326b9986 100644
--- a/core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
+++ b/core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
@@ -293,8 +291,10 @@ public class LoadBalancerConfigService {
         Set<InstanceGroup> knoxGatewayInstanceGroups = stack.getInstanceGroups().stream()
             .filter(ig -> knoxGatewayGroupNames.contains(ig.getGroupName()))
             .collect(Collectors.toSet());
-        validateKnoxForAzure(knoxGatewayInstanceGroups, stack.getCloudPlatform());
-        if (!knoxGatewayInstanceGroups.isEmpty()) {
+
+        if (AZURE.equalsIgnoreCase(stack.getCloudPlatform()) && knoxGatewayInstanceGroups.size() > 1) {
+            LOGGER.warn("For Azure load balancers, Knox must be defined in a single instance group. Load balancers will not be created.");
+        } else if (!knoxGatewayInstanceGroups.isEmpty()) {
             LOGGER.info("Knox gateway instance found; enabling Knox load balancer configuration.");
             knoxTargetGroup = new TargetGroup();
             knoxTargetGroup.setType(TargetGroupType.KNOX);
@@ -310,13 +310,6 @@ public class LoadBalancerConfigService {
         return Optional.ofNullable(knoxTargetGroup);
     }
 
-    private void validateKnoxForAzure(Set<InstanceGroup> knoxGatewayInstanceGroups, String cloudPlatform) {
-        if (AZURE.equalsIgnoreCase(cloudPlatform) && knoxGatewayInstanceGroups.size() > 1) {
-            LOGGER.warn("For Azure load balancers, Knox must be defined in a single instance group. Load balancers will not be created.");
-            knoxGatewayInstanceGroups.clear();
-        }
-    }
-
     private LoadBalancer createLoadBalancerIfNotExists(Set<LoadBalancer> loadBalancers, LoadBalancerType type, Stack stack) {
         LoadBalancer loadBalancer;
         Optional<LoadBalancer> existingLoadBalancer = loadBalancers.stream()
```

Reply:
@hreeve-cloudera How do you feel about throwing an error vs. skipping LB provisioning?

Reply:
I think in this case it would make sense to throw an error instead of silently failing. I'm leery of making all of the cases where we don't create a load balancer throw an error. We need to take a look at what cases may be valid for not creating an LB (I suspect there will be some QE cases that are), but we can do that as a separate effort. For now you can just make the change to throw an error in this particular case.

Reply:
I've thrown an exception now:
```java
if (AZURE.equalsIgnoreCase(stack.getCloudPlatform()) && knoxGatewayInstanceGroups.size() > 1) {
    throw new CloudbreakServiceException("For Azure load balancers, Knox must be defined in a single instance group.");
} else if (!knoxGatewayInstanceGroups.isEmpty()) {
```